library/kotlin/src/io/envoyproxy/envoymobile/StreamEmitter.kt

```diff
@@ -0,0 +1,52 @@
+package io.envoyproxy.envoymobile
+
+import java.nio.ByteBuffer
+
+interface StreamEmitter {
+  /**
+   * Identify if the emitter is still active and can perform communications with an associated stream.
+   *
+   * @return true if the stream is active.
+   */
+  fun isActive(): Boolean
+
+  /**
+   * For sending data to an associated stream.
+   *
+   * @param byteBuffer the byte buffer data to send to the stream.
+   * @throws IllegalStateException when the stream is not active
+   * @throws EnvoyException when there is an exception sending data.
+   * @return this stream emitter.
+   */
+  @Throws(EnvoyException::class)
+  fun sendData(byteBuffer: ByteBuffer): StreamEmitter
```

Comment:
How come this and metadata don't have an `end_stream` any more?

Reply:
Answered here: https://github.com/lyft/envoy-mobile/pull/265#discussion_r305954370

Reply:
We still need the ability to end the stream at any point, meaning as soon as you send headers/data/metadata, not only when you send trailers

Reply:
The approach we took here is that to close the stream, the consumer can call `close()` at any point. This can include trailers, or leave them empty (in which case none will be sent). So for example:

```
stream
  .sendData(data)
  .close() // or .close(trailers: myTrailers)
```

Reply:
This makes it explicit to the consumer when they're closing a stream, and gives them the option to include a set of trailers before the stream closes (as opposed to implicitly closing using a boolean flag)

Reply:
I'm good with this approach. I think it's probably more intuitive to people who aren't familiar with the underlying protocols, and it's both simple and valid to close a stream at any point with an empty `DATA` frame.

It also neatly solves the invalid chaining issue.

Reply:
Ok sounds good. As long as the three of you are aligned I'll get on board. I am a little concerned about the departure from the underlying API, but if this is more ergonomic at this layer, that makes sense.

Reply:
Yeah, it differ slightly but it doesn't necessarily hide any underlying functionality